### DISCUSSION TITLE
Can I change the current MediaStream?

### DISCUSSION THREAD

**Discussion by HighOnLemonz** [sipsorcery-org/sipsorcery#1253] [Q&A] [answered]
Hello,

Can I change the currently ongoing media stream?
I started the stream earlier and tried this at a later point:

//_peerConnection.removeTrack(_videoTrack);
_mediaFileSource = new FFmpegFileSource(_streams[requestStream.Current.AppId], true, null);
_mediaFileSource.RestrictFormats(x => x.Codec == VideoCodecsEnum.H264);
_videoTrack = new MediaStreamTrack(_mediaFileSource.GetVideoSourceFormats(), MediaStreamStatusEnum.SendOnly);
_peerConnection.addTrack(_videoTrack);
_mediaFileSource.OnVideoSourceEncodedSample += _peerConnection.SendVideo;

Removing the previous track would stop the stream altogether. Running the code as above will continue streaming the previously started stream. What am I doing wrong please? 

Renegociation is not needed because the 2 videos are of same resolution and use the same codec.

Thank you

**Comment by ha-ves** [comment]
If your `MediaStream` doesn't change (same codec, same specs), you can switch the event,

from:
```
_ffmpegVidOne.OnVideoSourceEncodedSample += PC.SendVideo
```

To
```
_ffmpegVidOne.OnVideoSourceEncodedSample -= PC.SendVideo
_ffmpegVidTwo.OnVideoSourceEncodedSample += PC.SendVideo
```

**Comment by HighOnLemonz** [comment]
It will always play the last track added(_peerConnection.addTrack) before the stream starts and cannot be modified afterwards. If I try your way it will work only for the added track and when I try to play the other track it will simply not send anything.

If it's working for you could you please provide a code sample? Thank you.

**Reply by ha-ves** [reply]
If you're going to go back to the previous video, then you can do it vice versa.

from:
```c#
_ffmpegVidTwo.OnVideoSourceEncodedSample += PC.SendVideo
```
To
```c#
_ffmpegVidTwo.OnVideoSourceEncodedSample -= PC.SendVideo
_ffmpegVidOne.OnVideoSourceEncodedSample += PC.SendVideo
```
---
Or, did you mean to do a multi-stream with two videos?

If you did mean Multi-track, then you'd need to renegotiate as needed for the additional `MediaStream` and use `PC.VideoStreamList[<index>].SendVideo` instead.


I can't promise an example for the time being though.

**Reply by HighOnLemonz** [reply]
Sorry for the misunderstanding. The client will wish to switch the streamed video. From my side I want to either "load" multiple tracks and then stream the one the client askes for or load the new track on the spot and start streaming it then. Only video streaming is required.

I assume that calling addTrack multiple times before the negotiation would get rid of the need to renegotiate so adding the videos beforehand should work, if that's the case.

**Reply by ha-ves** [reply, marked as answer]
All good 👍

>The client will wish to switch the streamed video. From my side, I want to either "load" multiple tracks and then stream the one the client askes for or load the new track on the spot and start streaming it then. Only video streaming is required.

This is possible, the receiving WebRTC implementation would also need to do the switching of the track being selected by the user in case of a single view.

If it's a browser, usually you will change the video element source to the selected MediaStream and then pause the other streams.

If it's a multi-view at once, then set the video element to the corresponding MediaStreams.

> I assume that calling addTrack multiple times before the negotiation would get rid of the need to renegotiate so adding the videos beforehand should work, if that's the case.

If you're creating multiple MediaStreamTrack (calling addTrack(s)) before negotiation, then it will all be included in the negotiation, so no need for renegotiation.
